developer_guides/virtualisation/virtualisation.rst

```diff
@@ -0,0 +1,360 @@
+to talk to the host they are running on. A typical Linux virtualisation
+implementation uses QEMU + KVM to run guest instances. In this case there are
+two ways to implement a VirtIO host: in the user space and in the kernel. If
+implemented in the user space, it is managed by QEMU. In the kernel space VirtIO
+host drivers use the Linux vhost API.
+
+It is also worth to note, that VirtIO guest drivers reside in respective driver
+subdirectories in the Linux kernel, e.g. drivers/net, drivers/block etc. On the
+contrary, all vhost drivers are located under drivers/vhost. That directory also
+contains vhost.c and vhost.h which are used by all vhost drivers.
+
+Every VirtIO driver requires explicit support in QEMU, which can be very minimal
+and simple in cases where all the functionality is implemented in VirtIO and
+vhost drivers, or it can actually implement a user-space counterpart to VirtIO
+drivers on the guest. In our case the SOF QEMU support code is rather minimal,
+it only adds support for guest reset reporting and guest identification to a set
+of standard and mostly empty device node and virtual queue management
+operations.
+
+2. SOF VirtIO design
+********************
+
+The SOF_ VirtIO design implements an interface to transport IPC (Inter-Process
+Communication) commands and audio data between host and guests. It re-uses the
+SOF-defined IPC API, also used by the native SOF driver to communicate with the
+SOF firmware, running on the audio DSP. This means, that the virtualisation is
+"applied" between the SOF driver and the DSP, i.e. IPC messages that the guest
+SOF driver would send to the DSP are instead sent to the guest, which then in
+the normal case forwards them to the DSP firmware, and then returns the response.
+
+Typically IPC message exchange between the driver and the DSP takes place over a
+mailbox and a doorbell. A mailbox is a shared memory region between DSP and
+host. It is used for IPC message payload. Doorbell is asynchronous signalling
+hardware between DSP and host. This mailbox / doorbell interface can be hardware
+specific, hence every hardware variant provides its own set of mailbox /
+doorbell handling operations. This virtualisation solution creates its own set
+of IPC operations to redirect IPC messages from the guest over virtual queues to
+the host.
+
+We use the same IPC message structures and we add 3 more VirtIO-specific IPC
+messages, which are never sent to the DSP, so the firmware doesn't need to
+support them. For the most part IPCs from guests are forwarded without change by
+the host driver to the DSP, only some of them are intercepted on the host. Some
+IPC messages trigger additional actions on the host before being forwarded to
+the DSP, others aren't sent to the DSP at all. See examples below for more
+details.
+
+2.1. Communication channels
+---------------------------
+
+The goal of the SOF virtualisation is to support differentiation in guest audio
```

Comment:
I understand what you mean by differentiation but that's an Intel buzzword.

maybe break in two objectives
- first goal is to enable streaming to/from audio endpoints, with the lowest possible overhead
- second goal is to allow a guest to control some of the processing on the DSP, while isolating it from other parts of the graph owned by the host or other VMs.